### ISSUE TITLE
Feature Request: Document states (open, closed), assigning document IDs and immutability

### ISSUE THREAD

**Issue by noseshimself** [dragotin/kraft#170]
(low priority; this will need a complete internal redesign to get the document into the database)

Currently documents are essigned their document ID upon creation and due to legal requirements become permanent, indestructible and should be (in the case of invoices, letters...) immutable.

This is not fitting everybody's workflow. It would be easier if documents could be modified in any way (including deleting them) until they are assigned a final/official document ID and up to this point be considered some kind of template. Only after explicitly locking the document the document should become a valid document and if it still needs changes there should be a function to turn the locked document into a correct reversal invoice ("Stornorechnung") {yes, this will need a different document template because it requires explicitly mentioning the cancelled document by ID).

**Comment by dragotin** [comment]
Yes, currently the document ID is assigned on the first save. In cases where the date is used in the doc identifier template that might even lead to wrong ids. 
This needs to change long term.
